[SPARK-36263][SQL][PYTHON] Add Dataframe.observation to PySpark

EnricoMi opened the pull request:
### What changes were proposed in this pull request?
With SPARK-34806 we can now easily add an equivalent for `Dataset.observe(Observation, Column, Column*)` to PySpark's `DataFrame` API.

### Why are the changes needed?
This further aligns the Python DataFrame API with Scala Dataset API.

### Does this PR introduce _any_ user-facing change?
Yes, it adds the `Observation` class and the `DataFrame.observe` method.

### How was this patch tested?
Adds test `test_observe` to `pyspark.sql.test.test_dataframe`.

EnricoMi committed:
Add Observation and DataFrame.observe to PySpark

EnricoMi committed:
Add python tests

EnricoMi committed:
Adjust to latest Observation API

EnricoMi committed:
Update docstrings

EnricoMi commented:
@cloud-fan here is a follow-up to #33422.

AmplabJenkins commented:
Can one of the admins verify this patch?

EnricoMi committed:
Simplify unit tests

EnricoMi committed:
Fix docstring example, move test_observe into pyspark.sql.test.test_dataframe

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,32 @@ def agg(self, *exprs):
+
+        :param observation: :class:`Observation` instance
+        :param exprs: aggregation expressions
+        :return: observed :class:`DataFrame`
```
Can you follow NumPy documentation style? (see also https://spark.apache.org/docs/latest/api/python/development/contributing.html#code-and-docstring-guide)

EnricoMi replied:
Done

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,100 @@
+        field_names = jrow.schema().fieldNames()
+        valuesScalaMap = jrow.getValuesMap(self._jvm.PythonUtils.toSeq(list(field_names)))
+        valuesJavaMap = self._jvm.scala.collection.JavaConversions.mapAsJavaMap(valuesScalaMap)
+        return Row(**valuesJavaMap)
```
it also has to have a `_test` method with `if __name__ == "__main__":` in order to enable doctests. Should also add this module into https://github.com/apache/spark/blob/master/dev/sparktestsupport/modules.py#L439

EnricoMi replied:
I was wondering how you are testing those code examples without defining `df` et al. Thanks for the pointers.

HyukjinKwon commented on python/pyspark/sql/tests/test_dataframe.py:
```diff
@@ -389,6 +391,32 @@ def test_extended_hint_types(self):
         self.assertEqual(1, logical_plan.toString().count("what"))
         self.assertEqual(3, logical_plan.toString().count("itworks"))
 
+    def test_observe(self):
```
Let's add a comment here, see "Pull Requests" at https://spark.apache.org/contributing.html

EnricoMi replied:
Done

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #141536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141536/testReport)** for PR 33484 at commit [`d192861`](https://github.com/apache/spark/commit/d19286134f43a4347e12de27d50cf5210b38fc99).

SparkQA commented:
**[Test build #141536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141536/testReport)** for PR 33484 at commit [`d192861`](https://github.com/apache/spark/commit/d19286134f43a4347e12de27d50cf5210b38fc99).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141536/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46054/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46054/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46054/

EnricoMi committed:
Review comments on docstring, python test setup

SparkQA commented:
**[Test build #141545 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141545/testReport)** for PR 33484 at commit [`fbde8d2`](https://github.com/apache/spark/commit/fbde8d2e01fa878f03c51e2117389abd4fa913fc).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46063/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46063/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46063/

SparkQA commented:
**[Test build #141545 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141545/testReport)** for PR 33484 at commit [`fbde8d2`](https://github.com/apache/spark/commit/fbde8d2e01fa878f03c51e2117389abd4fa913fc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141545/

EnricoMi commented:
>     * This patch **fails Spark unit tests**.

Unrelated `org.apache.spark.sql.kafka010.KafkaRelationSuiteV2.explicit offsets` failed with timeout.

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,40 @@ def agg(self, *exprs):
+
+        A user can retrieve the metrics by accessing `Observation.get`.
+
+        Example:
```
Let's format properly per https://numpydoc.readthedocs.io/en/latest/format.html#examples

EnricoMi replied:
Done

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,40 @@ def agg(self, *exprs):
+            >>> observed_df.count()
+            >>> metrics = observation.get
+
+        parameters
```
```suggestion
        Parameters
```

BTW, there's a certain order of these Parameters, Examples, etc. to follow.

EnricoMi replied:
I have also fixed a lowercase `Parameters` in another method. Fixed the order to Parameters, Returns, Examples.

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,40 @@ def agg(self, *exprs):
+        exprs : list of :class:`Column`
+            column expressions (:class:`Column`).
+
+        returns
```
```suggestion
        Returns
```

EnricoMi replied:
Fixed

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,146 @@
+
+    (failure_count, test_count) = doctest.testmod(
+        pyspark.sql.dataframe, globs=globs,
+        optionflags=doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE | doctest.REPORT_NDIFF)
```
I think we don't need to set all of the flags below:

```
        optionflags=doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE | doctest.REPORT_NDIFF)
```

EnricoMi replied:
done

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,146 @@
+
+    def _to_row(self, jrow):
+        field_names = jrow.schema().fieldNames()
+        valuesScalaMap = jrow.getValuesMap(self._jvm.PythonUtils.toSeq(list(field_names)))
```
Let's keep them in the under_score_naming whenever possible (except calling Py4J and user-facing API name)

EnricoMi replied:
fixed

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,146 @@
+        return DataFrame(observed_df, df.sql_ctx)
+
+    @property
+    def get(self) -> Row:
```
These type hints should be placed under a separate stub (.pyi) file for now.

EnricoMi replied:
done

HyukjinKwon commented on dev/sparktestsupport/modules.py:
```diff
@@ -437,6 +437,7 @@ def __hash__(self):
         "pyspark.sql.pandas.serializers",
         "pyspark.sql.pandas.typehints",
         "pyspark.sql.pandas.utils",
+        "pyspark.sql.observation",
```
let's put this in PySpark documentation as well https://github.com/apache/spark/blob/master/python/docs/source/reference/pyspark.sql.rst

EnricoMi replied:
so many magic places... done

EnricoMi committed:
Review comments on docstrings, documentation and typing

SparkQA commented:
**[Test build #141553 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141553/testReport)** for PR 33484 at commit [`7fc332d`](https://github.com/apache/spark/commit/7fc332dd0fafc406fe7a3728462ea514cbe77420).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46071/

SparkQA commented:
**[Test build #141559 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141559/testReport)** for PR 33484 at commit [`49e9b34`](https://github.com/apache/spark/commit/49e9b34d7719866b79d9fd129485ad1ce6209c1b).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46071/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46071/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46077/

SparkQA commented:
**[Test build #141553 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141553/testReport)** for PR 33484 at commit [`7fc332d`](https://github.com/apache/spark/commit/7fc332dd0fafc406fe7a3728462ea514cbe77420).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46077/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141553/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46077/

EnricoMi committed:
Fixing Observation examples

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #141559 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141559/testReport)** for PR 33484 at commit [`49e9b34`](https://github.com/apache/spark/commit/49e9b34d7719866b79d9fd129485ad1ce6209c1b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141559/

SparkQA commented:
**[Test build #141565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141565/testReport)** for PR 33484 at commit [`6d0a418`](https://github.com/apache/spark/commit/6d0a418ae2b3788a5d2fbf573ce992db18687faa).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46083/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46083/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46083/

SparkQA commented:
**[Test build #141565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141565/testReport)** for PR 33484 at commit [`6d0a418`](https://github.com/apache/spark/commit/6d0a418ae2b3788a5d2fbf573ce992db18687faa).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141565/

SparkQA commented:
**[Test build #141585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141585/testReport)** for PR 33484 at commit [`92d785b`](https://github.com/apache/spark/commit/92d785b6663105a0f35a7390dc2642495d1a75b2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46102/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46102/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46102/

SparkQA commented:
**[Test build #141585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141585/testReport)** for PR 33484 at commit [`92d785b`](https://github.com/apache/spark/commit/92d785b6663105a0f35a7390dc2642495d1a75b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141585/

EnricoMi commented:
@HyukjinKwon thanks for the input and all the pointer, everything looks green now.

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

HyukjinKwon commented on python/pyspark/sql/observation.pyi:
```diff
@@ -0,0 +1,31 @@
+
+class Observation:
+    def __init__(self, name: Optional[str] = ...): ...
+    def _on(self, df: DataFrame, *exprs: Column) -> DataFrame: ...
```
you can remove `_on` and `_to_row`. We'll only expose type hints for APIs only for now.

EnricoMi replied:
done

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+
+from pyspark.sql import column, Column, DataFrame, Row
+
+basestring = unicode = str
```
we dopped Python 2 so this alias is not needed anymore

EnricoMi replied:
removed

HyukjinKwon replied:
@EnricoMi seems like removal here is missed.

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,43 @@ def agg(self, *exprs):
+    def observe(self, observation, *exprs):
+        """Observe (named) metrics through an :class:`Observation` instance.
+
+        This method does not support streaming datasets.
```
Let's mode this to:

```
Notes
------
This method does not support streaming datasets.
```

EnricoMi replied:
moved

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,43 @@ def agg(self, *exprs):
+
+        This method does not support streaming datasets.
+
+        .. versionadded:: 3.3.0
```
Let's move this under "A user can retrieve ..."

EnricoMi replied:
moved

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,43 @@ def agg(self, *exprs):
+        Examples
+        --------
+        >>> from pyspark.sql.functions import col, count, lit, max
+        >>> from pyspark.sql.observation import Observation
```
Can you add the import of `Observation` to here: https://github.com/apache/spark/blob/master/python/pyspark/sql/__init__.py ?

So users can directly import it from `from pyspark.sql import Observation`. Let's fix the examples here as well

EnricoMi replied:
I like that, done

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -1830,6 +1830,43 @@ def agg(self, *exprs):
+
+        Examples
+        --------
+        >>> from pyspark.sql.functions import col, count, lit, max
```
Let's remove `count` and `max` here

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+        self._jvm = None
+        self._jo = None
+
+    def _on(self, df, *exprs):
```
can we keep the same signature as Scala side? e.g.) `_on(self, df, expr, *exprs)` to disallow empty expr

EnricoMi replied:
The method tests for empty exprs, but the Scala signature checks this on compile time, which is better.

I have aligned that with existing `agg` method, which has a single `exprs` and checks for non-emptiness.

Do you still say to mimic the Scala signature? Would you even say DataFrame.agg should do the same? I could draft another PR to fix that for agg if you let me. That shouldn't break python user code, right?

HyukjinKwon replied:
Hm, okay. let's leave it as is for now.

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+    Metrics are aggregation expressions, which are applied to the DataFrame while is is being
+    processed by an action.
+
+    .. versionadded:: 3.3.0
```
Let's also move this to the bottom of this description, and move `This class does not support streaming datasets` to `Notes` section.

EnricoMi replied:
both moved

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+        jrow = self._jo.get()
+        return self._to_row(jrow)
+
+    def _to_row(self, jrow):
```
@cloud-fan @EnricoMi @hvanhovell, shall we return a Map instead?

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+
+        Returns
+        -------
+        :class:`Row`
```
@cloud-fan @EnricoMi @hvanhovell, shall we return a Map instead?

EnricoMi replied:
I think if we move this to a Map, than it should be the same in Scala and Java. And there should be a getAsJavaMap then to avoid something like this in Java:

    JavaConverters.mapAsJavaMap(namedObservation.get())

HyukjinKwon replied:
Yeah, I think we might need to just return a map. I'll merge this PR as is for now but would you mind creating a followup PR to return a map please?

EnricoMi replied:
Yes, this would touch unrelated code and with a followup PR we can keep this possibly controversial change separated. Will do.

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+    import pyspark.sql.observation
+    globs = pyspark.sql.observation.__dict__.copy()
+    sc = SparkContext('local[4]', 'PythonTest')
+    globs['sc'] = sc
```
```suggestion
```

EnricoMi replied:
the self-contained example now needs `sc`, so I keep this

EnricoMi replied:
well, removed it as self-contained test now uses `spark`

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+    sc = SparkContext('local[4]', 'PythonTest')
+    globs['sc'] = sc
+    globs['spark'] = SparkSession(sc)
+    globs['df'] = sc.parallelize([(2, 'Alice'), (5, 'Bob')]) \
```
Can we show the DataFrame creation in the example codes? I know there are a lot of places like this already but I am thinking about changing it to show in the examples, and make the example self-contained.

EnricoMi replied:
I agree, within this file it should be self-contained. In `dataframe.py` it shouldn't be repeated.

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+        assert self._jo is None, "an Observation can be used with a DataFrame only once"
+
+        self._jvm = df._sc._jvm
+        self._jo = self._jvm.org.apache.spark.sql.Observation(self._name)
```
I think we don't need to create uuid in Python side. Just call `self._jvm.org.apache.spark.sql.Observation()` if name is not given.

EnricoMi replied:
yes, agreed, this is redundant

HyukjinKwon commented on python/pyspark/sql/observation.py:
```diff
@@ -0,0 +1,147 @@
+                          StructField('name', StringType())]))
+
+    (failure_count, test_count) = doctest.testmod(pyspark.sql.observation, globs=globs)
+    globs['sc'].stop()
```
```suggestion
    sc.stop()
```

EnricoMi replied:
the self-contained example now needs `sc`, so I keep this

HyukjinKwon reviewed (commented):
Looks good otherwise

SparkQA commented:
**[Test build #141603 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141603/testReport)** for PR 33484 at commit [`92d785b`](https://github.com/apache/spark/commit/92d785b6663105a0f35a7390dc2642495d1a75b2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46119/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46119/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46119/

SparkQA commented:
**[Test build #141603 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141603/testReport)** for PR 33484 at commit [`92d785b`](https://github.com/apache/spark/commit/92d785b6663105a0f35a7390dc2642495d1a75b2).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141603/

EnricoMi committed:
Another round of PR comments

- make Observation importable from pyspark.sql
- move note that streaming datasets are not supported into Notes section
- move versionadded to the end of the description
- no need to create a uuid in Python
- create example dataframe in observation.py example to be self-contained
- remove private methods from observation.pyi
- add tests for unnamed, empty name observations and observed aggregate columns type

EnricoMi commented:
@HyukjinKwon Thanks for the thorough comments above. Waiting for outcome on ["using Scala signature" comment](https://github.com/apache/spark/pull/33484#discussion_r676122947) and ["returning Map" comment](https://github.com/apache/spark/pull/33484#discussion_r676124454) before pushing above fixes.

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
@EnricoMi seems like you forgot to push the changes you made 😂

SparkQA commented:
**[Test build #141610 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141610/testReport)** for PR 33484 at commit [`92d785b`](https://github.com/apache/spark/commit/92d785b6663105a0f35a7390dc2642495d1a75b2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46127/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46127/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46127/

SparkQA commented:
**[Test build #141610 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141610/testReport)** for PR 33484 at commit [`92d785b`](https://github.com/apache/spark/commit/92d785b6663105a0f35a7390dc2642495d1a75b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141610/

EnricoMi commented:
> @EnricoMi seems like you forgot to push the changes you made joy

@HyukjinKwon seems like you haven't seen [my comment on that](https://github.com/apache/spark/pull/33484#issuecomment-886188794) :smile:. No worries, you have answered my last two questions. Will push now.

SparkQA commented:
**[Test build #141615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141615/testReport)** for PR 33484 at commit [`fa8c802`](https://github.com/apache/spark/commit/fa8c8026a8f62b9654a06e60cee2c618172c6870).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46132/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46132/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46132/

SparkQA commented:
**[Test build #141615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141615/testReport)** for PR 33484 at commit [`fa8c802`](https://github.com/apache/spark/commit/fa8c8026a8f62b9654a06e60cee2c618172c6870).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141615/

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #141634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141634/testReport)** for PR 33484 at commit [`fa8c802`](https://github.com/apache/spark/commit/fa8c8026a8f62b9654a06e60cee2c618172c6870).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46150/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46150/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46150/

SparkQA commented:
**[Test build #141634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141634/testReport)** for PR 33484 at commit [`fa8c802`](https://github.com/apache/spark/commit/fa8c8026a8f62b9654a06e60cee2c618172c6870).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141634/

EnricoMi commented:
>     * This patch **fails Spark unit tests**.

@HyukjinKwon These test failures seem to be transient and unrelated. This time `org.apache.spark.sql.kafka010.KafkaRelationSuiteV2.SPARK-30656: minPartitions` failed, last time it was `org.apache.spark.sql.execution.ui.AllExecutionsPageSuite.SPARK-27019: correctly display SQL page when event reordering happens`.

HyukjinKwon commented:
Can you rerun Github actions job? https://github.com/G-Research/spark/actions/runs/1066412255

cloud-fan commented:
retest this please

EnricoMi commented:
@HyukjinKwon I have rebased my branch to latest master, there are so many red jobs and every rerun they change.

SparkQA commented:
**[Test build #141704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141704/testReport)** for PR 33484 at commit [`90675b0`](https://github.com/apache/spark/commit/90675b0d300a299a047f97274e9d0814b2366a37).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46217/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46217/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46217/

SparkQA commented:
**[Test build #141704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141704/testReport)** for PR 33484 at commit [`90675b0`](https://github.com/apache/spark/commit/90675b0d300a299a047f97274e9d0814b2366a37).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141704/

EnricoMi commented:
Yay, all green! Quick, before something reruns :smile:.

cloud-fan commented:
thanks, merging to master!
